File: core/handoff.py
"""Agent Handoff — Task delegation between agents.

Inspired by OpenAI Agents SDK's handoff pattern, agents can delegate
tasks to specialized agents. The LLM decides when to hand off based
on the agent's handoff configuration.

Usage:
    from flowforge.core.handoff import Handoff, HandoffManager

    # Define handoffs for an agent
    handoffs = [
        Handoff(target="topic_agent", condition="research and topic selection"),
        Handoff(target="writing_agent", condition="article writing and editing"),
        Handoff(target="review_agent", condition="content review and quality check"),
    ]

    # Register with HandoffManager
    hm = HandoffManager(agent_registry=agent_registry)
    hm.register_handoffs("coordinator_agent", handoffs)

    # Execute handoff (typically called by the agent during execution)
    result = await hm.execute_handoff(
        source_agent="coordinator_agent",
        target_agent="topic_agent",
        task="Research trending AI topics",
        context={"persona": "tech"}
    )
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

from flowforge.core.agent_registry import AgentRegistry
from flowforge.core.base_agent import AgentInput, AgentOutput
from flowforge.core.tracing import get_logger

logger = get_logger("handoff")


class Handoff(BaseModel):
    """Defines a delegation from one agent to another.

    Attributes:
        target: Name of the target agent to delegate to.
        condition: Description of when this handoff should be used.
        description: Optional longer description of the handoff purpose.
        metadata: Arbitrary metadata for the handoff.
    """

    target: str
    condition: str
    description: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class HandoffManager:
    """Manages handoff routing between agents.

    The HandoffManager validates target agents, transfers task context,
    executes the target agent, and returns the result to the caller.
    It also generates prompts that tell the LLM about available handoffs.
    """

    def __init__(self, agent_registry: AgentRegistry) -> None:
        self._agent_registry = agent_registry
        self._handoffs: dict[str, list[Handoff]] = {}

    def register_handoffs(self, agent_name: str, handoffs: list[Handoff]) -> None:
        """Register handoffs for a given agent.

        Args:
            agent_name: The name of the source agent.
            handoffs: List of Handoff configurations.
        """
        existing = self._handoffs.get(agent_name, [])
        # Merge: add new handoffs that don't duplicate existing targets
        existing_targets = {h.target for h in existing}
        for h in handoffs:
            if h.target not in existing_targets:
                existing.append(h)
                existing_targets.add(h.target)
                logger.info(f"Registered handoff: {agent_name} -> {h.target} ({h.condition})")
            else:
                logger.debug(f"Handoff to '{h.target}' already exists for '{agent_name}', skipping")
        self._handoffs[agent_name] = existing

    def get_handoffs(self, agent_name: str) -> list[Handoff]:
        """Get all handoffs registered for a given agent.

        Args:
            agent_name: The name of the source agent.

        Returns:
            A list of Handoff configurations. Empty list if none registered.
        """
        return list(self._handoffs.get(agent_name, []))

    async def execute_handoff(
        self,
        source_agent: str,
        target_agent: str,
        task: str,
        context: dict[str, Any] | None = None,
    ) -> AgentOutput:
        """Execute a handoff from one agent to another.

        Validates the target agent exists, constructs an AgentInput
        with the task and context, executes the target agent, and
        returns the result.

        Args:
            source_agent: Name of the agent initiating the handoff.
            target_agent: Name of the target agent to delegate to.
            task: The task description to hand off.
            context: Additional context to pass to the target agent.

        Returns:
            An AgentOutput from the target agent's execution.

        Raises:
            ValueError: If the target agent is not registered.
            ValueError: If no handoff is configured from source to target.
        """
        context = context or {}

        # Validate handoff is configured
        handoffs = self.get_handoffs(source_agent)
        valid_targets = {h.target for h in handoffs}
        if target_agent not in valid_targets:
            raise ValueError(
                f"No handoff configured from '{source_agent}' to '{target_agent}'. "
                f"Available targets: {sorted(valid_targets)}"
            )

        # Validate target agent exists
        agent = self._agent_registry.get(target_agent)
        if agent is None:
            raise ValueError(
                f"Target agent '{target_agent}' not found in registry. "
                f"Registered agents: {self._agent_registry.list_agents()}"
            )

        logger.info(f"Handoff: {source_agent} -> {target_agent} | task: {task[:100]}")

        # Build input with task and context
        params: dict[str, Any] = {"task": task, **context}
        agent_input = AgentInput(params=params, state=context.get("state"))

        # Execute the target agent
        try:
            result = await agent.execute(agent_input)
            logger.info(f"Handoff complete: {source_agent} -> {target_agent}")
            return result
        except Exception as e:
            logger.error(f"Handoff execution failed ({source_agent} -> {target_agent}): {e}")
            return AgentOutput(result={"error": str(e), "handoff_from": source_agent})

    def get_handoff_prompt(self, agent_name: str) -> str:
        """Generate a prompt that tells the LLM about available handoffs.

        This prompt is injected into the agent's system prompt so the
        LLM knows when it can delegate tasks to other agents.

        Args:
            agent_name: The name of the agent to generate the prompt for.

        Returns:
            A string describing available handoffs, or empty string
            if no handoffs are configured.
        """
        handoffs = self.get_handoffs(agent_name)
        if not handoffs:
            return ""

        lines = [
            "You can delegate tasks to the following specialized agents:",
            "",
        ]
        for h in handoffs:
            line = f"- {h.target}: {h.condition}"
            if h.description:
                line += f" ({h.description})"
            lines.append(line)

        lines.append("")
        lines.append(
            "To delegate a task, indicate which agent should handle it "
            "and provide the task description."
        )

        return "\n".join(lines)

File: packages/core/state/src/handoff.ts
/**
 * handoff — Agent Handoff 任务交接（TS 重写自 `core/handoff.py`，F27）。
 *
 * 对齐 OpenAI Agents SDK handoff 模式：agent 可将任务委派给专业 agent。
 * 依赖通过 `AgentRegistryLike` 接口注入（Cordis 装配）。
 *
 * @module @flowforge/core-state
 */

/** 目标 agent 最小接口（执行输入返回输出）。 */
export interface HandoffAgent {
  execute(input: {
    params: Record<string, unknown>;
    state?: unknown;
  }): Promise<{ result: Record<string, unknown>; error?: string }>;
}

/** agent 注册表最小接口（get/list）。 */
export interface AgentRegistryLike {
  get(name: string): HandoffAgent | undefined;
  list(): string[];
}

/** 交接执行输出。 */
export interface HandoffOutput {
  result: Record<string, unknown>;
  error?: string;
}

/** 交接配置：从源 agent 委派到目标 agent。 */
export interface Handoff {
  /** 目标 agent 名称。 */
  target: string;
  /** 何时应使用此交接的描述。 */
  condition: string;
  /** 可选的交接目的长描述。 */
  description?: string;
  /** 任意元数据。 */
  metadata?: Record<string, unknown>;
}

/** 管理 agent 间交接路由。 */
export class HandoffManager {
  private readonly registry: AgentRegistryLike;
  private readonly handoffs = new Map<string, Handoff[]>();

  constructor(agentRegistry: AgentRegistryLike) {
    this.registry = agentRegistry;
  }

  /** 注册某源 agent 的交接配置（按 target 去重合并）。 */
  registerHandoffs(agentName: string, handoffs: Handoff[]): void {
    const existing = this.handoffs.get(agentName) ?? [];
    const existingTargets = new Set(existing.map((h) => h.target));
    for (const h of handoffs) {
      if (!existingTargets.has(h.target)) {
        existing.push(h);
        existingTargets.add(h.target);
      }
    }
    this.handoffs.set(agentName, existing);
  }

  /** 获取某源 agent 已注册的交接配置（无则空数组）。 */
  getHandoffs(agentName: string): Handoff[] {
    return [...(this.handoffs.get(agentName) ?? [])];
  }

  /**
   * 执行一次交接：校验配置与目标存在 → 构造输入 → 执行目标 agent。
   *
   * @throws 未配置交接 / 目标 agent 不存在时抛错。
   */
  async executeHandoff(
    sourceAgent: string,
    targetAgent: string,
    task: string,
    context: Record<string, unknown> = {},
  ): Promise<HandoffOutput> {
    // 校验交接已配置
    const validTargets = new Set(
      this.getHandoffs(sourceAgent).map((h) => h.target),
    );
    if (!validTargets.has(targetAgent)) {
      throw new Error(
        `No handoff configured from '${sourceAgent}' to '${targetAgent}'. ` +
          `Available targets: ${[...validTargets].sort().join(', ')}`,
      );
    }

    // 校验目标 agent 存在
    const agent = this.registry.get(targetAgent);
    if (agent === undefined) {
      throw new Error(
        `Target agent '${targetAgent}' not found in registry. ` +
          `Registered agents: ${this.registry.list().join(', ')}`,
      );
    }

    // 构造输入并执行
    const params = { task, ...context };
    try {
      return await agent.execute({
        params,
        state: context['state'],
      });
    } catch (e) {
      return {
        result: {
          error: e instanceof Error ? e.message : String(e),
          handoff_from: sourceAgent,
        },
      };
    }
  }

  /**
   * 生成告知 LLM 可用交接的提示词（注入系统提示）。
   * 无交接配置时返回空字符串。
   */
  getHandoffPrompt(agentName: string): string {
    const handoffs = this.getHandoffs(agentName);
    if (handoffs.length === 0) {
      return '';
    }
    const lines = [
      'You can delegate tasks to the following specialized agents:',
      '',
    ];
    for (const h of handoffs) {
      let line = `- ${h.target}: ${h.condition}`;
      if (h.description) {
        line += ` (${h.description})`;
      }
      lines.push(line);
    }
    lines.push(
      '',
      'To delegate a task, indicate which agent should handle it ' +
        'and provide the task description.',
    );
    return lines.join('\n');
  }
}
